pkg/ebpf/tracee.go

```diff
@@ -276,25 +271,62 @@ func New(cfg Config) (*Tracee, error) {
 		return nil, fmt.Errorf("error creating process tree: %v", err)
 	}
 
-	t.eventsToTrace = make(map[int32]bool, len(t.config.Filter.EventsToTrace))
+	t.events = make(map[int32]handledEvent, len(cfg.Filter.EventsToTrace))
+
+	// Set essential events that have to be loaded
+	t.setEventLoad(SysEnterEventID)
+	t.setEventLoad(SysExitEventID)
+	t.setEventLoad(SchedProcessForkEventID)
+	t.setEventLoad(SchedProcessExecEventID)
+	t.setEventLoad(SchedProcessExitEventID)
+	t.setEventLoad(CgroupMkdirEventID)
+	t.setEventSubmit(CgroupMkdirEventID)
+	t.setEventLoad(CgroupRmdirEventID)
+	t.setEventSubmit(CgroupRmdirEventID)
+
+	// Set events used to capture data
+	if t.config.Capture.Exec {
+		t.setEventLoad(SchedProcessExecEventID)
+		t.setEventSubmit(SchedProcessExecEventID)
+	}
+	if t.config.Capture.FileWrite {
+		t.setEventLoad(VfsWriteEventID)
+		t.setEventLoad(VfsWritevEventID)
+		t.setEventLoad(__KernelWriteEventID)
+	}
+	if t.config.Capture.Module {
+		t.setEventLoad(SecurityPostReadFileEventID)
+		t.setEventLoad(InitModuleEventID)
+	}
+	if t.config.Capture.Mem {
+		t.setEventLoad(MmapEventID)
+		t.setEventLoad(MprotectEventID)
+		t.setEventLoad(MemProtAlertEventID)
+	}
+	if t.config.Capture.NetIfaces != nil || cfg.Debug {
+		t.setEventLoad(SecuritySocketBindEventID)
+	}
+
+	// Events chosen by the user
 	for _, e := range t.config.Filter.EventsToTrace {
-		// Map value is true iff events requested by the user
-		t.eventsToTrace[e] = true
+		t.setEventLoad(e)
+		t.setEventSubmit(e)
+		t.setEventEmit(e)
 	}
 
 	initReq := RequiredInitValues{}
 	// Handles all essential events dependencies
-	for id, event := range EventsDefinitions {
-		if event.EssentialEvent || t.eventsToTrace[id] {
-			handleEventsDependencies(id, &initReq)
-		}
+	for id := range t.events {
+		t.handleEventsDependencies(id, &initReq)
 	}
-
-	// Compile final list of events to trace including essential events
-	for id, event := range EventsDefinitions {
-		// If an essential event was not requested by the user, set its map value to false
-		if event.EssentialEvent && !t.eventsToTrace[id] {
-			t.eventsToTrace[id] = false
+	if initReq.kallsyms {
+		t.kernelSymbols, err = helpers.NewKernelSymbolsMap()
+		if err != nil {
+			t.Close()
+			return nil, fmt.Errorf("error creating symbols map: %v", err)
+		}
+		if !initialization.ValidateKsymbolsTable(t.kernelSymbols) {
```

Comment:
don't you want to `t.Close()` here as well?

Reply:
If New() fails, tracee fails anyway, so no need to close here